File: lib/util/hrt_duration.dart
import 'dart:math';

import 'package:mona/data/model/date.dart';
import 'package:mona/i18n/translations.g.dart';

enum HrtDurationUnit { days, weeks, months, years }

class HrtDuration {
  final HrtDurationUnit unit;
  final int value;

  const HrtDuration(this.unit, this.value);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is HrtDuration && other.unit == unit && other.value == value;

  @override
  int get hashCode => Object.hash(unit, value);

  @override
  String toString() => 'HrtDuration($unit, $value)';
}

const int _weeksThresholdDays = 7;
const int _monthsThresholdDays = 90;

HrtDuration hrtDurationSince(Date start) {
  final now = Date.today();
  final days = start.differenceInDays(now);

  if (days < _weeksThresholdDays) {
    return HrtDuration(HrtDurationUnit.days, max(days, 1));
  }

  if (days < _monthsThresholdDays) {
    return HrtDuration(HrtDurationUnit.weeks, days ~/ 7);
  }

  final years = start.differenceInYears(now);
  if (years < 1) {
    return HrtDuration(HrtDurationUnit.months, start.differenceInMonths(now));
  }

  return HrtDuration(HrtDurationUnit.years, years);
}

String hrtDurationText(HrtDuration d) => switch (d.unit) {
      HrtDurationUnit.days => t.onHrtForDays(count: d.value),
      HrtDurationUnit.weeks => t.onHrtForWeeks(count: d.value),
      HrtDurationUnit.months => t.onHrtForMonths(count: d.value),
      HrtDurationUnit.years => t.onHrtForYears(count: d.value),
    };

String? hrtWidgetDurationText({
  required Date? firstTakenLocalDate,
  required bool intakeCounterEnabled,
}) =>
    firstTakenLocalDate == null || !intakeCounterEnabled
        ? null
        : hrtDurationText(hrtDurationSince(firstTakenLocalDate));

File: android/app/src/main/kotlin/com/deliacheminot/mona/HrtDuration.kt
package com.deliacheminot.mona

import java.time.LocalDate
import java.time.temporal.ChronoUnit

enum class HrtDurationUnit { DAYS, WEEKS, MONTHS, YEARS }

data class HrtDuration(val unit: HrtDurationUnit, val value: Int)

private const val WEEKS_THRESHOLD_DAYS = 7L
private const val MONTHS_THRESHOLD_DAYS = 90L

fun hrtDurationSince(startIso: String, today: LocalDate): HrtDuration {
    val start = LocalDate.parse(startIso)
    val days = ChronoUnit.DAYS.between(start, today).let { if (it < 0) -it else it }

    if (days < WEEKS_THRESHOLD_DAYS) {
        return HrtDuration(HrtDurationUnit.DAYS, maxOf(days, 1L).toInt())
    }
    if (days < MONTHS_THRESHOLD_DAYS) {
        return HrtDuration(HrtDurationUnit.WEEKS, (days / 7L).toInt())
    }
    val months = ChronoUnit.MONTHS.between(start, today).let { if (it < 0) -it else it }
    val years = months / 12L
    if (years < 1L) {
        return HrtDuration(HrtDurationUnit.MONTHS, months.toInt())
    }
    return HrtDuration(HrtDurationUnit.YEARS, years.toInt())
}
